pkg/virt-api/rest/subresource.go

```diff
@@ -520,6 +527,97 @@ func (app *SubresourceAPIApp) RestartVMRequestHandler(request *restful.Request,
 	response.WriteHeader(http.StatusAccepted)
 }
 
+func (app *SubresourceAPIApp) RenameVMRequestHandler(request *restful.Request, response *restful.Response) {
+	name := request.PathParameter("name")
+	namespace := request.PathParameter("namespace")
+
+	opts := &v1.RenameOptions{}
+
+	if request.Request.Body != nil {
+		defer request.Request.Body.Close()
+		err := yaml.NewYAMLOrJSONDecoder(request.Request.Body, 1024).Decode(opts)
+		switch err {
+		case io.EOF, nil:
+			break
+		default:
+			writeError(errors.NewBadRequest(fmt.Sprintf("Can not unmarshal Request body to struct, error: %s",
+				err)), response)
+			return
+		}
+	} else {
+		writeError(errors.NewBadRequest("Request with no body, a new name is expected as the request body"),
+			response)
+		return
+	}
+
+	if opts.NewName == "" {
+		writeError(errors.NewBadRequest("Please provide a new name for the VM"), response)
+		return
+	}
+
+	if name == opts.NewName {
+		writeError(errors.NewBadRequest("The VM's new name cannot be identical to the current name"), response)
+		return
+	}
+
+	// Make sure the VM is stopped and was not scheduled for renaming already
+	vm, statusErr := app.fetchVirtualMachine(name, namespace)
+
+	if statusErr != nil {
+		writeError(statusErr, response)
+		return
+	}
+
+	// Check for rename request on VM
+	for _, changeRequest := range vm.Status.StateChangeRequests {
+		if changeRequest.Action == v1.RenameRequest {
+			writeError(errors.NewConflict(v1.Resource("virtualmachine"),
+				name, fmt.Errorf("VM is already scheduled to be renamed")), response)
+			return
+		}
+	}
+
+	// Make sure VM is stopped
+	runningStatus, _ := vm.RunStrategy()
+
+	if runningStatus != v1.RunStrategyHalted {
+		writeError(errors.NewBadRequest("Renaming a running VM is not allowed"), response)
+		return
+	}
+
+	// Make sure a VM with the newName doesn't exist
+	_, statusErr = app.fetchVirtualMachine(opts.NewName, namespace)
+
+	if statusErr == nil {
+		writeError(errors.NewBadRequest("A VM with the new name already exists"), response)
+		return
+	} else if statusErr.ErrStatus.Code != http.StatusNotFound {
+		writeError(statusErr, response)
+		return
+	}
+
+	renameRequestJson, err := getChangeRequestJson(vm, v1.VirtualMachineStateChangeRequest{
+		Action: v1.RenameRequest,
+		Data: map[string]string{
+			"newName": opts.NewName,
+		},
+	})
+
+	if err != nil {
+		writeError(errors.NewInternalError(err), response)
+		return
+	}
+
+	_, err = app.virtCli.VirtualMachine(namespace).Patch(name, types.JSONPatchType, []byte(renameRequestJson))
```

Comment:
Per the resolved convo above:  I forget the details of how the different patch types work, but some (I think json patch) will replace lists rather than add elements to the list.  Have you confirmed (is there a test) where a rename request is added to a non empty list?

Reply:
I am using `getChangeRequestJson` (a few lines above), which prevents adding a `RenameRequest` to `Status.StateChangeRequests` if its not empty.
That is a valid scenario because I don't want to rename a VM that is scheduled to run because it might have an active VMI by the time renaming is done, and I don't want to rename a VM that is scheduled to stop for the same reason.
Because of that I think there is no need to add a `RenameRequest` to the list, only replace the list when its empty.